### PR TITLE
[stable/kiam] Add Helm chart for kiam

### LINKED ISSUE #5329: Add Helm chart for kiam
**Is this a request for help?**: No

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

Add a Helm chart for [`kiam`](https://github.com/uswitch/kiam).
This issue supersedes https://github.com/uswitch/kiam/issues/30.

### PR DESCRIPTION
**Pull request by ewbankkit**
**What this PR does / why we need it**: Adds an incubating Helm chart for [`kiam`](https://github.com/uswitch/kiam)

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #5329 

**Special notes for your reviewer**:
This PR supersedes https://github.com/uswitch/kiam/pull/53.

### CODE REVIEW

**Comment by timstoop** [comment]
Please add the `home` and `appVersion` keys to Chart.yaml.

**Comment by ewbankkit** [comment]
@timstoop Done.

**Comment by ewbankkit** [comment]
I will need to deal with the pre-requisite of [secrets for TLS](https://github.com/uswitch/kiam/blob/master/docs/TLS.md) and also update `NOTES.txt` to provide instructions.

**Comment by timstoop** [comment]
And provide a dummy one for the CI.

File: incubator/kiam/README.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+```console
+$ helm install incubator/kiam
+```
+
+## Introduction
+
+This chart bootstraps a [kiam](https://github.com/uswitch/kiam) deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
+
+## Prerequisites
+  - Kubernetes 1.4+ with Beta APIs enabled
```
Comment (unguiculus) [inline incubator/kiam/README.md]: 1.8+?

File: incubator/kiam/templates/agent-daemonset.yaml [hunk 2]
```diff
@@ -0,0 +1,112 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline incubator/kiam/templates/agent-daemonset.yaml]: apps/v1beta2

File: incubator/kiam/templates/agent-daemonset.yaml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+kind: DaemonSet
+metadata:
+  labels:
+    app: {{ template "kiam.name" . }}
+    chart: {{ template "kiam.chart" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "kiam.fullname" . }}-agent
+  namespace: {{ .Release.Namespace }}
```
Comment (unguiculus) [inline incubator/kiam/templates/agent-daemonset.yaml]: Remove namespaces from resources. The user can provide the namespace when they install the chart.

File: incubator/kiam/templates/server-daemonset.yaml [hunk 4]
```diff
@@ -0,0 +1,125 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline incubator/kiam/templates/server-daemonset.yaml]: apps/v1beta2

File: incubator/kiam/templates/server-daemonset.yaml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "kiam.fullname" . }}-server
+  namespace: {{ .Release.Namespace }}
+spec:
+  template:
+    metadata:
+      annotations:
+        prometheus.io/scrape: {{ quote .Values.server.prometheus.scrape }}
+        prometheus.io/port: {{ quote .Values.server.prometheus.port }}
```
Comment (unguiculus) [inline incubator/kiam/templates/server-daemonset.yaml]: Remove hard-coded annotations. Users that need them can configure them.

File: incubator/kiam/templates/server-daemonset.yaml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+              - --server-address-refresh=2s
+              - --timeout=5s
+            initialDelaySeconds: 3
+            periodSeconds: 10
+            timeoutSeconds: 10
+        {{- if .Values.server.resources }}
+          resources:
+{{ toYaml .Values.server.resources | indent 12 }}
+        {{- end }}
+{{- if and (eq .Capabilities.KubeVersion.Major "1") (gt .Capabilities.KubeVersion.Minor "5") }}
```
Comment (unguiculus) [inline incubator/kiam/templates/server-daemonset.yaml]: Remove if block.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please apply our RBAC best practices:

https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md

**Comment by ewbankkit** [comment]
I have fixed the problem with the E2E tests; I'll address the other review comments.

File: incubator/kiam/README.md (hunk 1 above)
Comment (ewbankkit) [reply incubator/kiam/README.md]: Yes; Fixed.

File: incubator/kiam/templates/agent-daemonset.yaml (hunk 2 above)
Comment (ewbankkit) [reply incubator/kiam/templates/agent-daemonset.yaml]: Fixed.

File: incubator/kiam/templates/server-daemonset.yaml (hunk 4 above)
Comment (ewbankkit) [reply incubator/kiam/templates/server-daemonset.yaml]: Fixed.

File: incubator/kiam/templates/agent-daemonset.yaml (hunk 3 above)
Comment (ewbankkit) [reply incubator/kiam/templates/agent-daemonset.yaml]: Fixed.

File: incubator/kiam/templates/server-daemonset.yaml (hunk 5 above)
Comment (ewbankkit) [reply incubator/kiam/templates/server-daemonset.yaml]: Fixed.

File: incubator/kiam/templates/server-daemonset.yaml (hunk 6 above)
Comment (ewbankkit) [reply incubator/kiam/templates/server-daemonset.yaml]: Fixed, plus the same in `incubator/kiam/templates/agent-daemonset.yaml`.

File: incubator/kiam/templates/server-daemonset.yaml (hunk 4 above)
Comment (ewbankkit) [reply incubator/kiam/templates/server-daemonset.yaml]: This caused the addition of

```yaml
  selector:
    matchLabels:
```

to `spec`.

**Comment by unguiculus** [comment]
Please have a look at our RBAC best practices again and apply them.

It is possible to specify custom values for CI. I'd like you to to do that. You should remove `useSampleData`. It invites users to use that even if they shouldn't. Add base64-encoded certs directly to a custom values file instead.

https://github.com/kubernetes/charts/blob/master/test/README.md#providing-custom-test-values

File: incubator/kiam/templates/server-serviceaccount.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{{- if .Values.serviceAccounts.server.create }}
```
Comment (unguiculus) [inline incubator/kiam/templates/server-serviceaccount.yaml]: Add `.Values.server.enabled` to the condition.

File: incubator/kiam/templates/agent-serviceaccount.yaml [hunk 8]
```diff
@@ -1,11 +1,12 @@
-{{- if .Values.rbac.create -}}
-kind: ServiceAccount
+{{- if .Values.serviceAccounts.agent.create }}
```
Comment (unguiculus) [inline incubator/kiam/templates/agent-serviceaccount.yaml]: Add `.Values.agent.enabled` to the condition.

File: incubator/kiam/templates/_helpers.tpl [hunk 9]
```diff
@@ -24,9 +24,67 @@ If release name contains chart name it will be used as a full name.
 {{- end -}}
 {{- end -}}
 
+{{/*
+Create a fully qualified agent name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+If release name contains chart name it will be used as a full name.
+*/}}
+{{- define "kiam.agent.fullname" -}}
+{{- if .Values.agent.fullnameOverride -}}
```
Comment (unguiculus) [inline incubator/kiam/templates/_helpers.tpl]: Do you need agent and server names to be configurable? Can you just have the standard fullname template that's base on the chart name and append suffixes for agent and server, similar to what you had before?

File: incubator/kiam/values.yaml [hunk 10]
```diff
@@ -128,20 +143,18 @@ server:
...
-    caFileName: ca.pem
-    certFileName: server.pem
-    keyFileName: server-key.pem
-    ## If true, use sample TLS data.
-    ##
-    useSampleData: true
+
+  ## Base64-encoded PEM values for server's CA certificate(s), certificate and private key
+  ##
+  tlsFiles:
```
Comment (unguiculus) [inline incubator/kiam/values.yaml]: Don't user `null`. Just leave them empty.

File: incubator/kiam/Chart.yaml [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+keywords:
+- kiam
+- aws
+- iam
+- security
+home: https://github.com/uswitch/kiam
+sources:
+- https://github.com/uswitch/kiam
+maintainers:
+- name: Kit Ewbank
```
Comment (unguiculus) [inline incubator/kiam/Chart.yaml]: Use Github username.

File: incubator/kiam/templates/server-clusterrolebinding.yaml [hunk 12]
```diff
@@ -0,0 +1,20 @@
+{{- if .Values.rbac.create -}}
```
Comment (unguiculus) [inline incubator/kiam/templates/server-clusterrolebinding.yaml]: Add `.Values.server.enabled` to the condition.

File: incubator/kiam/templates/server-clusterrole.yaml [hunk 13]
```diff
@@ -0,0 +1,22 @@
+{{- if .Values.rbac.create -}}
```
Comment (unguiculus) [inline incubator/kiam/templates/server-clusterrole.yaml]: Add `.Values.server.enabled` to the condition.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/kiam/templates/_helpers.tpl (hunk 9 above)
Comment (ewbankkit) [reply incubator/kiam/templates/_helpers.tpl]: Can do; I just tweaked based on the Prometheus chart.

File: incubator/kiam/templates/server-clusterrole.yaml (hunk 13 above)
Comment (ewbankkit) [reply incubator/kiam/templates/server-clusterrole.yaml]: 👍

File: incubator/kiam/templates/server-clusterrolebinding.yaml (hunk 12 above)
Comment (ewbankkit) [reply incubator/kiam/templates/server-clusterrolebinding.yaml]: 👍

File: incubator/kiam/templates/server-serviceaccount.yaml (hunk 7 above)
Comment (ewbankkit) [reply incubator/kiam/templates/server-serviceaccount.yaml]: 👍

File: incubator/kiam/templates/agent-serviceaccount.yaml (hunk 8 above)
Comment (ewbankkit) [reply incubator/kiam/templates/agent-serviceaccount.yaml]: 👍

File: incubator/kiam/Chart.yaml (hunk 11 above)
Comment (ewbankkit) [reply incubator/kiam/Chart.yaml]: 👍

File: incubator/kiam/values.yaml (hunk 10 above)
Comment (ewbankkit) [reply incubator/kiam/values.yaml]: 👍

File: incubator/kiam/templates/_helpers.tpl (hunk 9 above)
Comment (ewbankkit) [reply incubator/kiam/templates/_helpers.tpl]: 👍

**Comment by unguiculus** [comment]
I'd suggest you move the chart to stable directly.

**Comment by ewbankkit** [comment]
@unguiculus Is there anything additional I have to do to move directly to stable (other than move the code and reword the PR title)?

File: incubator/kiam/README.md [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+This chart bootstraps a [kiam](https://github.com/uswitch/kiam) deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
+
+## Prerequisites
+  - Kubernetes 1.8+ with Beta APIs enabled
+
+## Installing the Chart
+
+In order for the chart to configure kiam correctly during the installation process you should have created and installed TLS certificates and private keys as described [here](https://github.com/uswitch/kiam/blob/master/docs/TLS.md).
+
+> **Tip**: The `hosts` field in the kiam server certificate should include the value _release-name_-server:_server-service-port_, e.g. `my-release-server:443`
```
Comment (luispollo) [inline incubator/kiam/README.md]: @ewbankkit I was banging my head against the wall for a few hours trying to figure an error I was seeing in the agent logs, which boiled down to the hosts in the server cert. I would like to suggest the following change:

> **Tip**: The `hosts` field in the kiam server certificate must include the value `<release-name>-kiam-server:<server-service-port>`, e.g. `my-release-kiam-server:443`
>
> If you don't include the exact hostname used by the kiam agent to connect to the server, you'll see a warning (which is really an error) in the agent logs similar to the following, and your pods will fail to obtain credentials:
>
> `{"level":"warning","msg":"error finding role for pod: rpc error: code = Unavailable desc = there is no connection available","pod.ip":"100.120.0.2","time":"2018-05-24T04:11:25Z"}`

(Note the added `kiam-` part added after `my-release` and before `server`)

Comment (ewbankkit) [reply incubator/kiam/README.md]: @luispollo Thanks, I'll add that.

**Comment by luispollo** [comment]
@ewbankkit @unguiculus Any chance this will get merged/published as a stable chart in the next couple of days? Thanks!

**Comment by unguiculus** [comment]
Please move it to stable. Rename the PR and make sure references in readme etc. are updated.

**Comment by geota** [comment]
@ewbankkit  tested this on a newly created KOPS cluster. Followed the README and I was able to get it installed and the services up and running.

However, when testing a pod - I am seeing this error in the kiam-server logs and the pod is falling back to the default node role.

```RequestError: send request failed\ncaused by: Post https://sts.amazonaws.com/: x509: failed to load system roots and no roots provided```

Maybe the cert bundle is missing from the kiam-server image?

```apk update && apk add ca-certificates``` ?

**Comment by ewbankkit** [comment]
@ungulculus OK. I’m away from the keyboard for a week. Will make those changes next week.

**Comment by ewbankkit** [comment]
@geota Yes, on certain Linux flavors you will need to mount CA certs from the host. I will provide an example once I get back.

**Comment by ewbankkit** [comment]
@geota Added the TLS certs `hostPath` values for Amazon Linux/RHEL to the comments for `extraHostPathMounts` in `values.yaml`.

**Comment by ewbankkit** [comment]
@unguiculus OK, I think I have done all the steps to move straight to _stable_.

**Comment by timstoop** [comment]
/lgtm
Nice addition :)

File: stable/kiam/templates/server-daemonset.yaml [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "kiam.fullname" . }}-server
+spec:
+  selector:
+    matchLabels:
+      app: {{ template "kiam.name" . }}
+      component: "{{ .Values.server.name }}"
+      release: {{ .Release.Name }}
+      {{- if .Values.server.podLabels }}
```
Comment (unguiculus) [inline stable/kiam/templates/server-daemonset.yaml]: Additional labels should not be part of the selector.

File: stable/kiam/templates/server-service.yaml [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "kiam.fullname" . }}-server
+spec:
+  clusterIP: None
+  selector:
+    app: {{ template "kiam.name" . }}
+    component: "{{ .Values.server.name }}"
+    release: {{ .Release.Name }}
+    {{- if .Values.server.podLabels }}
```
Comment (unguiculus) [inline stable/kiam/templates/server-service.yaml]: Additional labels should not be part of the selector.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/kiam/templates/server-daemonset.yaml (hunk 15 above)
Comment (ewbankkit) [reply stable/kiam/templates/server-daemonset.yaml]: Fixed.

File: stable/kiam/templates/server-service.yaml (hunk 16 above)
Comment (ewbankkit) [reply stable/kiam/templates/server-service.yaml]: Fixed.

**Review by unguiculus — CHANGES_REQUESTED** [review]
One more thing. Please use a consistent list indentation style throughout the chart. Either use zero or two spaces. Don't mix.

**Comment by ewbankkit** [comment]
OK, standardized on 2 spaces for list indentation.

**Review by unguiculus — APPROVED** [review]

**Pull request merged**